### PR TITLE
Add copy Commit ID button in commits list

### LINKED ISSUE #13816: How to copy commit id in repo commit page？
Please see the picture below. I can directly copy the commit id through this button in Github. Does gitea support this function?
image

### PR DESCRIPTION
**Pull request by velengel**
Resolve:
* https://github.com/go-gitea/gitea/issues/13816

Screenshots:

### CODE REVIEW

**Comment by lunny** [comment]
Could you post a screenshot?

**Comment by Gusted** [comment]
This seems to be a variant of #17416?

**Comment by asSqr** [comment]
Yes, we implement the copy procedure in #17416. I'm sorry. Here is a screenshot:

copy clipboard

**Comment by wxiaoguang** [comment]
And more information: we have a better clipboard support now (see the clipboard.js), so most `data-xxx` and `onclick` are not needed. There are some examples for clipboard usage (search `clipboard-text=` to see)

**Comment by wxiaoguang** [comment]
And more more information: this UI has some edge cases, so you should check all the cases to make sure the UI is correct.

See the comment https://github.com/go-gitea/gitea/pull/17416#issuecomment-950283525

**Comment by asSqr** [comment]
Thank you for your advice. Let me check it.

**Comment by silverwind** [comment]
You only need `data-clipboard-text`, all other attributes can be removed, as well as the localization strings (we should not have so many variations, `Copied` popup text is fine).

File: templates/repo/commits_list.tmpl [hunk 1]
```diff
@@ -55,6 +55,17 @@
...
+							{{$commitID := .ID.String}}
+							<span class="clipboard"
+								data-success="{{$.i18n.Tr "repo.commits.copy_sha_success"}}"
+								data-error="{{$.i18n.Tr "repo.commits.copy_sha_error"}}"
+								data-content="{{$.i18n.Tr "repo.commits.copy_sha_text"}}"
+								data-variation="inverted tiny"
+								data-clipboard-text="{{.ID}}"
+								onclick="'{{$commitID}}'.select(); document.execCommand('copy');"
+							>
+								{{svg "octicon-paste"}}
```
Comment (silverwind) [inline templates/repo/commits_list.tmpl]:
```suggestion
								{{svg "octicon-copy"}}
```

**Comment by asSqr** [comment]
I fixed in response to your reviews. The edge case (commit with signature) seems correctly handled:

スクリーンショット 2021-11-23 0 04 43

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1052,6 +1052,9 @@ commits.search.tooltip = You can prefix keywords with "author:", "committer:", "
 commits.find = Search
 commits.search_all = All Branches
 commits.author = Author
+commits.copy_sha_text = Copy
+commits.copy_sha_success = Commit ID has been copied
+commits.copy_sha_error = Use ⌘C or Ctrl-C to copy
```
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
```

No longer needed

**Comment by silverwind** [comment]
Maybe put it on the left side of the SHA and add `ui button` class to style it like a button? (Same height as SHA box)

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (asSqr) [reply options/locale/locale_en-US.ini]: Thank you

**Comment by asSqr** [comment]
> Maybe put it on the left side of the SHA and add `ui button` class to style it like a button? (Same height as SHA box)

Yes, sir. Let me fix it.

**Comment by asSqr** [comment]
I fixed:

スクリーンショット 2021-11-23 1 00 52

**Comment by wxiaoguang** [comment]
That's better. I tune a little, please help to confirm.


image

image

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> Button is too large based on that screenshot, it needs to match the SHA button height. Also, it should use a `<button>` tag to be semantically correct and accessible.

@silverwind  Can you edit on this PR directly to continue to improve? This features have been tried for at least 3 times, I think such a trivial feature shouldn't block that long time.

**Comment by silverwind** [comment]
Someone push this patch, I don't have permission to do it.

image


````diff
diff --git a/templates/repo/commits_list.tmpl b/templates/repo/commits_list.tmpl
index e61450ad6..4f3134f98 100644
--- a/templates/repo/commits_list.tmpl
+++ b/templates/repo/commits_list.tmpl
@@ -22,9 +22,10 @@
 								{{avatarByEmail .Author.Email .Author.Name 28 "mr-2"}}
 								{{$userName}}
 							{{end}}
 						</td>
-						<td class="sha">
+						<td class="sha df">
+							<a class="ui label copy-commit-sha df ac" data-clipboard-text="{{.ID}}">{{svg "octicon-copy" 14}}</a>
 							{{$class := "ui sha label"}}
 							{{if .Signature}}
 								{{$class = (printf "%s%s" $class " isSigned")}}
 								{{if .Verification.Verified}}
@@ -45,9 +46,8 @@
 								<a href="{{$.RepoLink}}/commit/{{.ID}}" rel="nofollow" class="{{$class}}">
 							{{else}}
 								<span class="{{$class}}">
 							{{end}}
-								<span data-clipboard-text="{{.ID}}">{{svg "octicon-copy"}}</span>
 								<span class="shortsha">{{ShortSha .ID.String}}</span>
 								{{if .Signature}}
 									{{template "repo/shabox_badge" dict "root" $ "verification" .Verification}}
 								{{end}}
diff --git a/web_src/less/_base.less b/web_src/less/_base.less
index c09f3a2bd..ea0be2d36 100644
--- a/web_src/less/_base.less
+++ b/web_src/less/_base.less
@@ -984,8 +984,13 @@ a.ui.card:hover,
     font-weight: normal;
     margin: 0 6px;
   }

+  .copy-commit-sha {
+    border: 1px solid var(--color-light-border);
+    margin-right: 3px;
+  }
+
   .button.truncate {
     display: inline-block;
     max-width: 100%;
     overflow: hidden;
@@ -1451,8 +1456,13 @@ a.ui.label:hover {
   background: var(--color-hover);
   color: var(--color-text);
 }

+.ui.labels a.label:hover,
+a.ui.label:hover {
+  border-color: var(--color-light-border);
+}
+
 .ui.label > .detail .icons {
   margin-right: .25em;
 }
```

**Comment by asSqr** [comment]
Thank you. I have pushed it.

**Review by silverwind — APPROVED** [review]

**Pull request merged**